hermes/probe/monitor.go

```diff
@@ -12,9 +12,155 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 //
+// Authors:
+// - Alicja Kwiecinska (kwiecinskaa@google.com) github: alicjakwie
+// - Evan Spendlove, GitHub: evanSpendlove.
+//
 // Monitor implements a probe that monitors a storage system using the
 // Hermes algorithm.
 
+// Package probe implements the probe that Hermes uses to monitor
+// a storage system.
+package probe
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/google/cloudprober/logger"
+	"github.com/google/cloudprober/metrics"
+	"github.com/google/cloudprober/probes/options"
+
+	metricpb "github.com/google/cloudprober/metrics/proto"
+	probepb "github.com/googleinterns/step224-2020/config/proto"
+)
+
 // MonitorProbe holds aggregate information about all probe runs, per-target.
 // It also holds the config and options used to initialise the probe.
-type MonitorProbe struct{}
+type MonitorProbe struct {
+	name    string
+	config  *probepb.HermesProbeDef
+	targets []*probepb.Target
+	opts    *options.Options
+
+	// probeOpLatency is used to record latency values with distinct labels
+	// per probe operation per target.
+	// Recommended usage: append(probeOpLatency[target][probeOp], newMetric)
+	probeOpLatency map[string]map[string][]*metrics.EventMetrics
+	// apiCallLatency is used to record latency values with distinct labels
+	// per API call per target.
+	// Recommended usage: append(apiCallLatency[target][apiCall], newMetric)
+	apiCallLatency map[string]map[string][]*metrics.EventMetrics
+	logger         *logger.Logger
+}
+
+// Init initializes the probe with the given parameters.
+// This is a required method to implement the cloudprober.Probes.Probe interface.
+func (mp *MonitorProbe) Init(name string, opts *options.Options) error {
+	conf, ok := opts.ProbeConf.(*probepb.HermesProbeDef) // Casting ProbeConf interface to HermesProbeDef type
+	if !ok {
+		return fmt.Errorf("invalid argument: opts.ProbeConf is not of type *probepb.HermesProbeDef")
+	}
+	mp.name = name
+	mp.config = conf
+	mp.targets = mp.config.GetTargets()
+	mp.opts = opts
+	mp.probeOpLatency = make(map[string]map[string][]*metrics.EventMetrics)
+	mp.apiCallLatency = make(map[string]map[string][]*metrics.EventMetrics)
+	mp.logger = opts.Logger
+
+	return nil
+}
+
+// Start runs the probe indefinitely, unless cancelled, at the configured interval.
+// Probe metrics will be sent via the metricChan at the beginning of the next probe run.
+// Arguments:
+//	- ctx: context provided for cancelling probe.
+//	- metricChan: bidirectional channel used for sending metrics to be surfaced.
```

Comment:
The channel passed by Cloudprober is bidirectional, but I realise now that we can define the channel arg to be unidirectional, and it will still work. We won't ever need to receive data via the channel.